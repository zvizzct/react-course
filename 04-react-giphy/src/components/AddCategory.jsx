import { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({ onNewCategory }) => {
  const [inputValue, setInputValue] = useState('');

  const onInputChange = (e) => {
    setInputValue(e.target.value);
  };

  const onSubmit = (e) => {
    e.preventDefault();
    const newInputValue = inputValue.trim();
    if (newInputValue.length <= 1) return;
    onNewCategory(newInputValue);
    setInputValue('');
  };

  return (
    <form onSubmit={onSubmit} aria-label="form">
      <input
        type="text"
        placeholder="Search all the GIFs and Stickers"
        value={inputValue}
        onChange={onInputChange}
      />
    </form>
  );
};

AddCategory.propTypes = {
  onNewCategory: PropTypes.func.isRequired,
};
