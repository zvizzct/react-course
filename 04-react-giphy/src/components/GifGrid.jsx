import { useEffect, useState } from 'react';
import { getGifs } from '../helpers/getGifs';

export const GifGrid = ({ category }) => {
  const [images, setImages] = useState([]);

  const getImages = async () => {
    const newImages = await getGifs(category);
    setImages(newImages);
  };
  useEffect(() => {
    getImages();
  }, []);

  console.log(images);
  return (
    <>
      <h3>{category}</h3>
      <ol>
        {images.map(({ id, url }) => (
          <li key={id}>
            <img src={url}></img>
          </li>
        ))}
      </ol>
    </>
  );
};
