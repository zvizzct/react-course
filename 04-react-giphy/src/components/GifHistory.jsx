export const GifHistory = ({ history, onSelectCategory }) => {
  const handleButtonHistory = (e) => {
    const newCat = e.target.innerText;
    onSelectCategory(newCat);
  };
  return (
    <div className="section-history">
      {history
        ? history.map((category) => (
            <button
              className="btn-history"
              key={category}
              onClick={handleButtonHistory}
            >
              {category}
            </button>
          ))
        : null}
    </div>
  );
};
