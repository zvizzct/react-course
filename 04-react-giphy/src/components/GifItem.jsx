import { useState } from 'react';
import PropTypes from 'prop-types';

export const GifItem = ({ title, url }) => {
  const [copied, setCopied] = useState(false);
  const copyClipBoard = async () => {
    await navigator.clipboard.writeText(url);
    setCopied(true);
    setTimeout(() => {
      setCopied(false);
    }, 1000);
  };

  return (
    <div className="card" onClick={copyClipBoard}>
      {copied && (
        <div
          className="copied-message"
          style={{
            position: 'absolute',
            top: 10,
            left: 10,
            zIndex: 1,
            color: 'white',
            fontWeight: 'bold',
          }}
        >
          Copied to clipboard
        </div>
      )}
      {copied ? (
        <img src={url} alt={title} style={{ filter: `brightness(50%)` }}></img>
      ) : (
        <img src={url} alt={title}></img>
      )}
    </div>
    // </li>
  );
};

GifItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
};
