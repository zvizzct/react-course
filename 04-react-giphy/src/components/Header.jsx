import { useState, useEffect } from 'react';

export const Header = () => {
  const [darkTheme, setDarkTheme] = useState(true);

  const toggleTheme = () => {
    setDarkTheme(!darkTheme);
  };

  useEffect(() => {
    document.body.className = darkTheme ? 'dark-theme' : 'light-theme';
  }, [darkTheme]);

  return (
    <>
      <header>
        <h1>Gif Mania</h1>
        <p onClick={toggleTheme} style={{ cursor: 'pointer' }}>
          {!darkTheme ? '🌙' : '☀️'}
        </p>
      </header>
    </>
  );
};
