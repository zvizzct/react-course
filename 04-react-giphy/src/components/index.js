export * from './AddCategory';
export * from './GifGrid';
export * from './GifItem';
export * from './Header';
export * from './GifHistory';
