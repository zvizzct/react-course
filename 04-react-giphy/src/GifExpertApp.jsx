import { useState } from 'react';
import { AddCategory, GifGrid, Header, GifHistory } from './components';
export const GifExpertApp = () => {
  const [categories, setCategories] = useState(['The Office']);
  const [history, setHistory] = useState([]);

  const onSelectCategory = (selectedCat) => {
    setCategories([selectedCat]);
  };
  const onAddCategory = (newCat) => {
    if (categories.includes(newCat)) return;
    setCategories((prevCategories) => [newCat]);
    if (history && history.includes(newCat)) return;
    setHistory((prevHistory) => [newCat, ...prevHistory]);
  };

  return (
    <div className="container">
      <Header />
      <AddCategory onNewCategory={onAddCategory} />
      <GifHistory history={history} onSelectCategory={onSelectCategory} />
      {categories.map((category) => (
        <GifGrid key={category} category={category} />
      ))}
    </div>
  );
};
