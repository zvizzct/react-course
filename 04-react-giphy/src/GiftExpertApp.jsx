import { useState } from 'react';
import { AddCategory, GifGrid, Header, GifHistory } from './components';
export const GiftExpertApp = () => {
  const [categories, setCategories] = useState([]);
  const [history, setHistory] = useState([]);

  const onSelectCategory = (selectedCat) => {
    setCategories([selectedCat]);
  };
  const onAddCategory = (newCat) => {
    if (categories.includes(newCat)) return;
    if (history && history.includes(newCat)) return;
    setCategories((prevCategories) => [newCat]);
    setHistory((prevHistory) => [newCat, ...prevHistory]);
    console.log(newCat, categories, history);
  };

  return (
    <div className="container">
      <Header />
      <AddCategory onNewCategory={onAddCategory} />
      <GifHistory history={history} onSelectCategory={onSelectCategory} />
      {categories.map((category) => (
        <GifGrid key={category} category={category} />
      ))}
    </div>
  );
};
