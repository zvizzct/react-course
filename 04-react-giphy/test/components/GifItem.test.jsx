import {
  fireEvent,
  render,
  screen,
  act,
  getByText,
  getByRole,
} from '@testing-library/react';
import { GifItem } from '../../src/components';

describe('Testing GifItem', () => {
  const title = 'Un titulo';
  const url = 'https://localhost/algo.jpg';
  const msg = 'copied to clipboard';
  test('Debe de hacer match con el snapshot  ', () => {
    const { container } = render(<GifItem title={title} url={url} />);
    expect(container).toMatchSnapshot();
  });
  test('debe de mostrar la imagen con el url y el alt indicado', () => {
    render(<GifItem title={title} url={url} />);
    // expect(screen.getByRole('img').src).toBe(url);
    // expect(screen.getByRole('img').alt).toBe(title);

    const { src, alt } = screen.getByRole('img');
    expect(src).toBe(url);
    expect(alt).toBe(title);
  });
  test('displays copied message when clicked', async () => {
    render(<GifItem title={title} url={url} />);

    // Mock clipboard API
    global.navigator.clipboard = {
      writeText: jest.fn(),
    };

    await act(async () => {
      fireEvent.click(screen.getByRole('img'));
    });

    expect(global.navigator.clipboard.writeText).toHaveBeenCalledWith(url);
  });
});
