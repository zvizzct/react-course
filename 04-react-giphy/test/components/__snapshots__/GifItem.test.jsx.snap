// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing GifItem Debe de hacer match con el snapshot   1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Un titulo"
      src="https://localhost/algo.jpg"
    />
  </div>
</div>
`;
